library/core/src/num/int_sqrt.rs

```diff
@@ -0,0 +1,285 @@
+        let (mut r, overflow) =
+            ((u << QUARTER_BITS) | (lo & LOWEST_QUARTER_1_BITS)).overflowing_sub(q * q);
+        if overflow {
+            r = r.wrapping_add(2 * s - 1);
+            s -= 1;
+        }
+        (s, r)
+    }};
+}
+
+/// Generates the last stage of the computation before denormalization.
+macro_rules! last_stage {
+    ($ty:ty, $n:ident, $s:ident, $r:ident) => {{
+        // SAFETY: Inform the optimizer that `$s` is nonzero. This will allow
+        // it to avoid generating code to handle division-by-zero panics in the
+        // division below.
+        //
+        // See the proof in the `middle_stage` macro above.
+        unsafe { core::hint::assert_unchecked($s != 0) };
+
+        const HALF_BITS: u32 = <$ty>::BITS >> 1;
+        const QUARTER_BITS: u32 = <$ty>::BITS >> 2;
+        const LOWER_HALF_1_BITS: $ty = (1 << HALF_BITS) - 1;
+
+        let lo = $n & LOWER_HALF_1_BITS;
+        let numerator = (($r as $ty) << QUARTER_BITS) | (lo >> QUARTER_BITS);
+        let denominator = ($s as $ty) << 1;
+
+        let q = numerator / denominator;
+        let mut s = ($s << QUARTER_BITS) as $ty + q;
+        let (s_squared, overflow) = s.overflowing_mul(s);
+        if overflow || s_squared > $n {
+            s -= 1;
+        }
+        s
+    }};
+}
+
+/// Generates the stages of the computation between normalization and
+/// denormalization for [`u16`](prim@u16).
+macro_rules! u16_stages {
+    ($n:ident) => {{
+        let (s, r) = first_stage!(16, $n);
+        last_stage!(u16, $n, s, r)
+    }};
+}
+
+/// Generates the stages of the computation between normalization and
+/// denormalization for [`u32`](prim@u32).
+macro_rules! u32_stages {
+    ($n:ident) => {{
+        let (s, r) = first_stage!(32, $n);
+        let (s, r) = middle_stage!(32, u16, $n, s, r);
+        last_stage!(u32, $n, s, r)
+    }};
+}
+
+/// Generates the stages of the computation between normalization and
+/// denormalization for [`u64`](prim@u64).
+macro_rules! u64_stages {
+    ($n:ident) => {{
+        let (s, r) = first_stage!(64, $n);
+        let (s, r) = middle_stage!(64, u16, $n, s, r);
+        let (s, r) = middle_stage!(64, u32, $n, s, r);
+        last_stage!(u64, $n, s, r)
+    }};
+}
+
+/// Generates the stages of the computation between normalization and
+/// denormalization for [`u128`](prim@u128).
+macro_rules! u128_stages {
+    ($n:ident) => {{
+        let (s, r) = first_stage!(128, $n);
+        let (s, r) = middle_stage!(128, u16, $n, s, r);
+        let (s, r) = middle_stage!(128, u32, $n, s, r);
+        let (s, r) = middle_stage!(128, u64, $n, s, r);
+        last_stage!(u128, $n, s, r)
+    }};
+}
+
+unsigned_fn!(u16, u8, u16_stages);
+unsigned_fn!(u32, u16, u32_stages);
+unsigned_fn!(u64, u32, u64_stages);
+unsigned_fn!(u128, u64, u128_stages);
+
+/// Instantiate this panic logic once, rather than for all the isqrt methods
+/// on every single primitive type.
+#[cold]
```

Comment:
I did it that way because `int_log10` does that:

https://github.com/rust-lang/rust/blob/5601d14249818d952da612fec481b7af3ed03a39/library/core/src/num/int_log10.rs#L142-L148

My best guess as to why is that it might consider `panic!` cold but it won't consider `panic_for_nonpositive_argument` itself cold, and that might affect code generation, but I'm not sure. Should I remove it?

Reply:
I think that this usually makes sense if there is more in the function. But in this case the expansion is

```rust
#[cold]
pub const fn panic_for_nonpositive_argument() -> ! {
        ::core::panicking::panic_fmt(format_args!("argument of integer logarithm must be positive"));
}
```

`format_args` has no overhead, and `panic_fmt` is already cold. So I was thinking that by default, with no `#[cold]` on `panic_for_nonpositive_argument`, LLVM would inline `panic_for_nonpositive_argument`'s overhead (which is nothing, just a wasted call - it should be inlined). But then adding `#[cold]` could encourage outlining both `panic_for_nonpositive_argument` and `panic_fmt` in some circumstances. Which just means a tiny bit of wasted code in the panicking path.

Microoptimization obviously, I'm sure LLVM would inline the single call overhead in asm opts despite `#[cold]`.

(ChaiTRex this doesn't need to affect your code, I'm just figuring out my understanding)

Reply:
> I love the GH username by the way

Thanks!